### DISCUSSION TITLE
Your Emails Rock, and I'd Love To Bring this to .NET as a Pet Project

### DISCUSSION THREAD

**Discussion by XMLDiva** [bootstrap-email/bootstrap-email#194] [Kind Words]
I used to create emails over a decade ago and I haaaated it. I remember when I first found BootStrap Email about a year ago and was surprised no one's done it sooner. So thank you for creating this with such care, and even more thanks for the concept of converting bootstrap to correctly-janky email format.

I'm thinking of to porting this to .net as a pet project for myself (I need dynamic content emails shared that can be done without internet connection)  If I ever end up pushing this into .net's package manager, I'll make sure it's clear yours is the original inspiration but the .net is a completely unrelated (and name it differntly) so you don't get issues accidentally created.

**Comment by stuyam** [comment]
Since bootstrap email supports compilation via the CLI, I wonder if there is a way it could be used to power support in other projects like .NET. I would guess the system would need ruby installed, but maybe it can be packaged as a binary and create .NET bindings to work with it. Thoughts?

**Reply by stuyam** [reply]
Also thanks for the kind words, really appreciate the support ❤️

**Reply by XMLDiva** [reply]
After looking into it more, Ruby & .NET have not been playing nice for awhile (there was a Ruby engine in .net....12 years ago lol ). I think for just starting for what I'm doing, I think CLI will do. If I get too many blocks and really feel I need to go full nilly, I may revisit the full rewrite later.

**Reply by taylorchasewhite** [reply]
@XMLDiva is your consumption of this project from .NET open source somewhere that I can take a look at it? I too am looking to use this from a .NET hosted application.

**Comment by w8tcha** [comment]
> I'm thinking of to porting this to .net as a pet project for myself (I need dynamic content emails shared that can be done without internet connection) If I ever end up pushing this into .net's package manager, I'll make sure it's clear yours is the original inspiration but the .net is a completely unrelated (and name it differntly) so you don't get issues accidentally created.

I love this project, and use it in my .net forum software to create the email templates. I managed to convert the project to .NET 7/8 and convert it as library + cli. 

https://github.com/w8tcha/BootstrapEmail.Net

It exist as nuget package so it can be easily integrated in a .net project.

**Comment by JohnCampionJr** [comment]
We must have been working on this at the same time :)

Here's my attempt:
https://github.com/UnDotNet/BootstrapEmail

**Reply by w8tcha** [reply]
nice i added benchmarks tests to see which solution performs better...

```

BenchmarkDotNet v0.13.10, Windows 11 (10.0.22635.2771)
12th Gen Intel Core i9-12900K, 1 CPU, 24 logical and 16 physical cores
.NET SDK 8.0.100
  [Host]   : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  ShortRun : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1  
WarmupCount=3  

```
| Method                  | Data          | Mean     | Error     | StdDev   | Ratio | RatioSD | Gen0       | Gen1       | Gen2      | Allocated | Alloc Ratio |
|------------------------ |-------------- |---------:|----------:|---------:|------:|--------:|-----------:|-----------:|----------:|----------:|------------:|
| **UnDotNet.BootstrapEmail** | **alert**         | **58.07 ms** | **21.729 ms** | **1.191 ms** |  **1.00** |    **0.00** | **11000.0000** | **10000.0000** | **9000.0000** |  **64.09 MB** |        **1.00** |
| BootstrapEmail.Net      | alert         | 59.44 ms |  4.233 ms | 0.232 ms |  1.02 |    0.02 |  4111.1111 |  2444.4444 | 1555.5556 |  54.83 MB |        0.86 |
|                         |               |          |           |          |       |         |            |            |           |           |             |
| **UnDotNet.BootstrapEmail** | **basic**         | **49.86 ms** |  **3.847 ms** | **0.211 ms** |  **1.00** |    **0.00** | **10600.0000** |  **9800.0000** | **9100.0000** |  **60.12 MB** |        **1.00** |
| BootstrapEmail.Net      | basic         | 59.43 ms | 14.352 ms | 0.787 ms |  1.19 |    0.02 |  3777.7778 |  2222.2222 | 1111.1111 |  54.83 MB |        0.91 |
|                         |               |          |           |          |       |         |            |            |           |           |             |
| **UnDotNet.BootstrapEmail** | **border-radius** | **64.20 ms** |  **5.160 ms** | **0.283 ms** |  **1.00** |    **0.00** | **11000.0000** | **10000.0000** | **9000.0000** |  **71.61 MB** |        **1.00** |
| BootstrapEmail.Net      | border-radius | 58.97 ms | 17.146 ms | 0.940 ms |  0.92 |    0.01 |  3888.8889 |  2222.2222 | 1222.2222 |  54.83 MB |        0.77 |
|                         |               |          |           |          |       |         |            |            |           |           |             |
| **UnDotNet.BootstrapEmail** | **images**        | **50.21 ms** |  **8.722 ms** | **0.478 ms** |  **1.00** |    **0.00** | **10600.0000** |  **9800.0000** | **9100.0000** |  **59.94 MB** |        **1.00** |
| BootstrapEmail.Net      | images        | 60.89 ms | 55.176 ms | 3.024 ms |  1.21 |    0.07 |  3857.1429 |  2285.7143 | 1285.7143 |  54.83 MB |        0.91 |
|                         |               |          |           |          |       |         |            |            |           |           |             |
| **UnDotNet.BootstrapEmail** | **integration**   | **85.74 ms** | **78.173 ms** | **4.285 ms** |  **1.00** |    **0.00** | **10333.3333** |  **9000.0000** | **7333.3333** |  **81.49 MB** |        **1.00** |
| BootstrapEmail.Net      | integration   | 62.44 ms | 21.416 ms | 1.174 ms |  0.73 |    0.04 |  4000.0000 |  2444.4444 | 1333.3333 |  54.83 MB |        0.67 |

**Reply by JohnCampionJr** [reply]
Hi, very cool.  Your integration test is compiling a tiny bit of html while mine is doing the whole integration string.  Can you fix and update the results?

I think the big difference is you're compiling the SASS which makes total sense for a CLI use.  I am just using precompiled CSS because I want to auto generate these in another project, not via CLI, so that SASS will never change.
```
{
        var bsEmail = new BootstrapEmail(new ConfigStore());

        return bsEmail.Compile("<a href='#' class='btn btn-primary'>Some Button</a>", string.Empty, InputType.String);
    }
```

**Reply by w8tcha** [reply]
I fixed the test

> think the big difference is you're compiling the SASS which makes total sense for a CLI use. I am just using precompiled CSS because I want to auto generate these in another project, not via CLI, so that SASS will never change.

Like the original project the sass files are only compiled once then there are cached (stored).
